From 47d19c8beae75668b562328ce73fa93a43d4de58 Mon Sep 17 00:00:00 2001
From: Zahari Doychev <zdoychev@maxlinear.com>
Date: Fri, 16 Apr 2021 09:26:21 +0200
Subject: [PATCH] dp: mib: fix memory leak in counters update

PONRTSYS-8859: The locally allocate memory in update_dev_counters is
now also freed on initial update of the counters.
---
 drivers/net/datapath/mib_counters/dp_mib_counters.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/drivers/net/datapath/mib_counters/dp_mib_counters.c b/drivers/net/datapath/mib_counters/dp_mib_counters.c
index 3fe3f4fe3943..0f053b528b4d 100644
--- a/drivers/net/datapath/mib_counters/dp_mib_counters.c
+++ b/drivers/net/datapath/mib_counters/dp_mib_counters.c
@@ -327,7 +327,7 @@ static void update_dev_counters(struct mib_counters *cnt, u8 initial_update)
 		if (rmon)
 			cnt->last.rx_ovfl_err =	rmon->cnt[RMON_RX_OVERFLOW];
 		write_unlock(&cnt->cnt_lock);
-		return;
+		goto free_exit;
 	}
 
 	/* update 64 bits wide counters based on last read values */
@@ -373,6 +373,7 @@ static void update_dev_counters(struct mib_counters *cnt, u8 initial_update)
 		cnt->last.rx_ovfl_err =	rmon->cnt[RMON_RX_OVERFLOW];
 	write_unlock(&cnt->cnt_lock);
 
+free_exit:
 	devm_kfree(&g_dp_dev->dev, tmp_counters);
 	if (rmon)
 		devm_kfree(&g_dp_dev->dev, rmon);
